### ISSUE TITLE
The sole C# example code does not compile and is invalid.

### ISSUE THREAD

**Issue by matthewacme** [MicrosoftDocs/windows-uwp#5022]
### Type of issue

Code doesn't work

### Feedback

in the section
https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/applifecycle/applifecycle-instancing#redirection-without-blocking

you supply a C# Static method to handle a non-blocking STA redirection. The example as it is provided in the documentation will not work or compile

The line
`keyInstance.Activated += OnActivated;`
will not work or compile, it causes a [Compiler Error CS0120](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0120?f1url=%3FappId%3Droslyn%26k%3Dk(CS0120))

the entire else block of 
```
else
{
	isRedirect = true;

	// Ensure we don't block the STA, by doing the redirect operation
	// in another thread, and using an event to signal when it has completed.
	redirectEventHandle = CreateEvent( IntPtr.Zero , true , false , null );
	if ( redirectEventHandle != IntPtr.Zero )
	{
		Task.Run( () =>
		{
			keyInstance.RedirectActivationToAsync( args ).AsTask().Wait();
			SetEvent( redirectEventHandle );
		} );
		uint CWMO_DEFAULT = 0;
		uint INFINITE = 0xFFFFFFFF;
		_ = CoWaitForMultipleObjects(
				CWMO_DEFAULT , INFINITE , 1 ,
				new IntPtr[] { redirectEventHandle } , out uint handleIndex );
	}
}
```
also will not compile

The `redirectEventHandle` variable has not be defined or declared
Neither the `CreateEvent` or `SetEvent` methods are provided in the example

The same is also true for the `CoWaitForMultipleObjects` method.

In short, your example code is GIGO

### Page URL

https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/applifecycle/applifecycle-instancing

### Content source URL

https://github.com/MicrosoftDocs/windows-dev-docs/blob/docs/hub/apps/windows-app-sdk/applifecycle/applifecycle-instancing.md

### Author

@Karl-Bridge-Microsoft

### Document Id

d362e3da-c88c-a4d7-c51d-dd9d98487c2a

**Comment by matthewacme** [comment]
here is the code corrected as best I can to get it to work (sort of) (see the note in the Static class where the CoWaitForMultipleObjects function is called)

I got rid of the static on the code that attempts to assign the event handler , and added a utility class for the COM calls.

```
	private bool DecideRedirection( string key )
	{
		bool isRedirect = false;

		// Find out what kind of activation this is.
		AppActivationArguments args = AppInstance.GetCurrent().GetActivatedEventArgs();
		ExtendedActivationKind kind = args.Kind;
		if ( kind == ExtendedActivationKind.Launch )
		{
			try
			{
				AppInstance keyInstance = AppInstance.FindOrRegisterForKey( key );

				// If we successfully registered the file name, we must be the
				// only instance running that was activated for this file.
				if ( keyInstance.IsCurrent )
				{
					// Hook up the Activated event, to allow for this instance of the app
					// getting reactivated as a result of multi-instance redirection.
					keyInstance.Activated += OnKeyInstanceActivated;
				}
				else
				{
					isRedirect = true;

					Utility.HandleWaiterInterop waiter = new HandleWaiterInterop();

					// Ensure we don't block the STA, by doing the redirect operation
					// in another thread, and using an event to signal when it has completed.
					var redirectEventHandle = HandleWaiterInterop.CreateEvent( IntPtr.Zero , true , false , null );
					if ( redirectEventHandle != IntPtr.Zero )
					{
						Task.Run( () =>
						{
							keyInstance.RedirectActivationToAsync( args ).AsTask().Wait();
							HandleWaiterInterop.SetEvent( redirectEventHandle );
						} );

						HandleWaiterInterop.WaitForHandles( new nint[] { redirectEventHandle } );
					}
				}
			}
			catch ( Exception ex )
			{
				Debug.WriteLine( $"Error getting instance information: {ex.Message}" );
			}
		}

		return isRedirect;
	}
```

and to support the com calls, I included this utility class

```
public class HandleWaiterInterop
{
	// Import the CoWaitForMultipleObjects function from combase.dll
	[DllImport( "combase.dll" , SetLastError = true )]
	public static extern int CoWaitForMultipleObjects(
		 uint dwFlags ,
		 uint dwTimeout ,
		 uint cHandles ,
		 IntPtr pHandles ,
		 out uint lpdwIndex
	);

	[DllImport( "kernel32.dll" , SetLastError = true )]
	public static extern IntPtr CreateEvent(
		IntPtr lpEventAttributes ,
		bool bManualReset ,
		bool bInitialState ,
		string lpName
	);

	[DllImport( "kernel32.dll" , SetLastError = true )]
	public static extern bool SetEvent( IntPtr hEvent );

	// Wait for multiple handles using CoWaitForMultipleObjects
	public static void WaitForHandles( IntPtr[] handles )
	{
		uint index;
		try
		{

			// NOTE: according to MS documentation at
			// https://github.com/MicrosoftDocs/windows-dev-docs/blob/docs/hub/apps/windows-app-sdk/applifecycle/applifecycle-instancing.md
			// this should work... but it doesn't it throws an error with a memory violation.
			// at least it does inside Visual Studio when run in debug, so could be VS's fault

			int result = CoWaitForMultipleObjects( 0 ,  // Flags (0 for default behavior)
															  ( uint )0xFFFFFFFF ,  // Timeout in milliseconds
															  ( uint )handles.Length ,
															  handles[ 0 ] ,
															  out index );

			if ( result == 0 )
			{
				Console.WriteLine( $"Handle at index {index} signaled." );
			}
			else
			{
				Console.WriteLine( "Wait failed or timed out." );
			}
		}
		catch ( Exception ex )
		{
			// we crashed, so do nothing
		}
	}
}
```

**Comment by matthewacme** [comment]
Mind you when the CoWaitForMultipleObjects function is called the following error is raised, so something is still not right.

But since I only see this error in Visual Studio and when run in release mode, it appears to work... maybe it just doesn't like doing Debug with the COM calls in Visual Studio.

System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

**Comment by matthewacme** [comment]
Okay, so I referenced the code in 

https://github.com/microsoft/WindowsAppSDK-Samples/blob/62316b1f7db932477d4a3811afb0bdaf91ae614b/Samples/AppLifecycle/Instancing/cs2/cs-winui-packaged/CsWinUiDesktopInstancing/CsWinUiDesktopInstancing/Program.cs#L151

and modified my wait procedure to the following (and was able to delete the Static utility class that I was using to pull in the COM calls

```
	private bool DecideRedirection( string key )
	{
		bool isRedirect = false;

		// Find out what kind of activation this is.
		AppActivationArguments args = AppInstance.GetCurrent().GetActivatedEventArgs();
		ExtendedActivationKind kind = args.Kind;
		if ( kind == ExtendedActivationKind.Launch )
		{
			try
			{
				AppInstance keyInstance = AppInstance.FindOrRegisterForKey( key );

				// If we successfully registered the file name, we must be the
				// only instance running that was activated for this file.
				if ( keyInstance.IsCurrent )
				{
					// Hook up the Activated event, to allow for this instance of the app
					// getting reactivated as a result of multi-instance redirection.
					keyInstance.Activated += OnKeyInstanceActivated;
				}
				else
				{
					isRedirect = true;

					//Utility.HandleWaiterInterop waiter = new HandleWaiterInterop();

					// Ensure we don't block the STA, by doing the redirect operation
					// in another thread, and using an event to signal when it has completed.
					var redirectSemphore = new Semaphore( 0 , 1 );
					Task.Run( () =>
					{
						keyInstance.RedirectActivationToAsync( args ).AsTask().Wait();
						redirectSemphore.Release();
					} );
					redirectSemphore.WaitOne();

				}
			}
			catch ( Exception ex )
			{
				Debug.WriteLine( $"Error getting instance information: {ex.Message}" );
			}
		}

		return isRedirect;
	}
```

and that seems to work with no errors.

**Comment by alvinashcraft** [comment]
Thanks for submitting your feedback and we apologize for the delayed response. I see that updates have not been made since the creation of this Issue, so I went ahead and created an internal work item to incorporate your suggestions. The timeline for this may vary, so I will go ahead and close this and circle back when the changes are live.

Though the Issue will be marked as Closed, we will monitor any replies that come afterwards so treat this like an open Issue. Many thanks!
